ships/0004-build-env-vars.md

```diff
@@ -0,0 +1,241 @@
+
+The `Build` and `BuildRun` APIs will be extended with a `spec.env` field.
+This is where a developer declares that they are appending environment variables to a build step.
+`env` will have an array of Kubernetes envrionment variables that will be appended to every step of the build strategy.
+
+Referring to the nodejs buildpacks story above, a developer who wants their build to install `dev` dependencies would do the following in their `Build` or `BuildRun`:
+
+```yaml
+spec:
+  sources:
+  - name: git
+    type: Git
+    git:
+      uri: https://github.com/sclorg/nodejs-ex.git
+  strategy:
+    kind: BuildStrategy
+    name: buildpacks-v3
+  env:
+  - name: NPM_INCLUDE
+    value: dev
+```
+
+Environment variables will always be appended to the environment variables declared in the build strategy's steps.
+If a `BuildRun` is started and either the referenced `Build` or the `BuildRun` has an environment variable name that collides with any step in the build strategy, the `BuildRun` should fail.
+A `Build` could reference environment variable names that collide with the build strategy, as it would be hard to guarantee that a `Build` does not have bad enviroment variable names over time.
+
+The environment variable API will take advantage of Kubernete's core `EnvVar` interface, which allows environment variable data to be sourced from a Secret or ConfigMap.
+
+A `BuildRun` can override the environment variables specified in a `Build` object.
+
+#### CLI Enhancements
+
+The `shp build run` command should be extended to allow environment variables to be passed in using the `-e` and `--env` flag.
+This flag can be specified more than once, and the value should use the `<key>=<value>` syntax.
+For example, to set the `NPM_INCLUDE` environment variable when invoking a run of the `buildpacks-nodejs` build, use the following:
+
+```shell
+$ shp build run buildpacks-nodejs -e NPM_INCLUDE=dev
+```
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+TBD
+
+### Release Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+TBD
+
+#### Upgrade Strategy [if necessary]
+
+TBD
+
+### Risks and Mitigations
+
+There is a risk that developers will want an environment variable to be set in one build step, but unset or restored to a default value in a different build step.
+This could be addressed in a future enhancemenent via separate API field (ex: `excludeFromSteps`).
+Users who are comfortable modifying build strategies can work around this issue by writing their own strategies with the desired environment variables fixed or exposed as a parameter.
+
+## Drawbacks
+
+The primary drawback is that build strategy authors need to provide fixed or default values for environment variables they want to control.
+A strategy author cannot simply declare that an environment variable is "off limits" - they must declare when a controlled env var is used and what value it should have.
```

Comment:
👍 This is a concern I've had, but we can start with what we have in this proposal. 

Also, when should we *really* care about off-limit environment variables ?  When a user uses an environment variable to break out of the sandbox of a container. However, a secure container platform wouldn't let you do that if the pod isn't being run with elevated privileges/capabilities.

Reply:
[CVE-2021-20182](https://access.redhat.com/security/cve/CVE-2021-20182) is a truly compelling case for a build author to declare an environment variable off limits. Setting the `BUILDAH_ISOLATION` env var to `chroot` AND running with the `privileged: true` security context exposes devices on the node (storage, network, etc.).